### PR TITLE
Bugfix/rfc 2451 rerebalance tests

### LINKED ISSUE #64412: `re_rebalance_coherence` doesn't properly account for fundamental types
As part of an effort to stabilize the `re_rebalance_coherence` RFC (#63599), I was creating tests trying to get more coverage of all possible scenarios, and I found that we don't seem to be handling fundamental types in the way that the RFC specifies.

In particular this test is presently accepted and yet ought to be rejected:

```rust
#![feature(re_rebalance_coherence)]

extern crate coherence_lib as lib;
use lib::*;
use std::rc::Rc;

struct Local;

impl<T> Remote1<Local> for Box<T> { /* should error */ }
impl<T> Remote1<Local> for &T { /* should error */ }

fn main() {}
```

I'll be adding this test (along with some others) in a PR soon with "FIXME" comments.

### PR DESCRIPTION
**Pull request by weiznich**
r? @nikomatsakis 

Fixes #64412 
Depends/Contains on #64414 

cc #55437 and #63599

### CODE REVIEW

**Comment by nikomatsakis** [comment]
Thanks @weiznich! This looks good to me. I think we might want to add some more tests that show what happens when fundamental is used in a few more positions. Something like:

- `impl<T> Trait<Box<T>, Local> for u32 { }` -- expect error
- `impl<T> Trait<Box<T>> for Local { }` -- expect OK
- `impl<T> Trait<Local> for Vec<Box<T>> { }` -- expect OK, just as it is ok for `Vec<T>`

Do you think you can add such tests? Feel free to add any further combinations that seem worth covering -- I see no reason not to "tile the space" here with tests in every combination we can think of.

**Comment by weiznich** [comment]
@nikomatsakis  I've added some more tests.

**Review by nikomatsakis — APPROVED** [review]
awesome, thanks!

**Comment by nikomatsakis** [comment]
r=me once CI is green :)

**Comment by weiznich** [comment]
@nikomatsakis CI is green.

**Comment by Centril** [comment]
Failed in https://github.com/rust-lang/rust/pull/64647#issuecomment-533766271, @bors r-

**Comment by weiznich** [comment]
@Centril I didn't touch the file producing those warnings. Should I just fix them or is there any specific reasons why this happens now?

**Comment by Centril** [comment]
@weiznich not sure why it happened exactly, but if you know how to fix it, then great :)

**Comment by weiznich** [comment]
@Centril Hopefully that commit does fix the warnings.

**Comment by nikomatsakis** [comment]
@weiznich I think we need to rename tests like `ui\coherence\impl[t]-foreign[local, fundamental[t]]-for-foreign.rs` to not have spaces. (i.e., `ui\coherence\impl[t]-foreign[local,fundamental[t]]-for-foreign.rs`).

**Comment by weiznich** [comment]
@nikomatsakis Done

**Comment by Centril** [comment]
Failed in https://github.com/rust-lang/rust/pull/64733#issuecomment-534667832, @bors r- 

Alex noted on Discord that this might be path-length or concurrency related (I'm guessing the former).

**Comment by weiznich** [comment]
@Centril Just to be sure: So making the test file path names shorter may fix the issue? If so, does anyone know the specific limit?

**Comment by Centril** [comment]
@weiznich Emphasis on "may" -- but worth a shot. ;)

**Comment by ecstatic-morse** [comment]
It's not the filename length, the limit is 255 characters. The line
``` 
= note: D:/a/1/s/mingw64/bin/../lib/gcc/x86_64-w64-mingw32/6.3.0/../../../../x86_64-w64-mingw32/bin/ld.exe: cannot find fundamental[t]]-for-foreign\auxiliary\coherence_lib.dll.lib: No such file or directory
```
suggests that someone is treating `[x,y]` as an escape sequence.

**Comment by ecstatic-morse** [comment]
@weiznich I think the problem is the comma. Try switching to a different separator. Maybe a semicolon?

**Comment by weiznich** [comment]
So let's try this version :shrug:

**Comment by ecstatic-morse** [comment]
@weiznich There should be an even number of renames. You still have a comma somewhere.

File: src/test/ui/coherence/impl[t]-foreign[foreign[t],local]-for-foreign.rs [hunk 1]
```diff
@@ -0,0 +1,14 @@
+#![feature(re_rebalance_coherence)]
```
Comment (Centril) [inline src/test/ui/coherence/impl[t]-foreign[foreign[t],local]-for-foreign.rs]: here is the comma.

**Comment by weiznich** [comment]
@ecstatic-morse Not necessarily, because not all of the tests are compile fail tests.

**Comment by weiznich** [comment]
@Centril Next try?

**Pull request merged**